content/en/blog/_posts/2021-12-07-dual-stack-networking-ga.md

```diff
@@ -0,0 +1,49 @@
+---
+layout: blog
+title: 'Kubernetes 1.23: Dual-stack IPv4/IPv6 Networking Reaches GA'
+date: 2021-12-07
+slug: dual-stack-networking-ga
+---
+
+**Author:** Bridget Kromhout (Microsoft)
+
+"When will Kubernetes have IPv6?" This question has been asked with increasing frequency ever since alpha support for IPv6 was first added in k8s v1.9. While Kubernetes has supported IPv6-only clusters since v1.18, migration from IPv4 to IPv6 was not yet possible at that point. At long last, [dual-stack IPv4/IPv6 networking](https://github.com/kubernetes/enhancements/tree/master/keps/sig-network/563-dual-stack/) has reached general availability (GA) in Kubernetes v1.23.
+
+What does dual-stack networking mean for you? Let’s take a look…
+
+
+## Service API updates
+
+[Services](/docs/concepts/services-networking/service/) were single-stack before 1.20, so using both IP families meant creating one Service per IP family. The user experience was simplified in 1.20, when Services were re-implemented to allow both IP families, meaning a single Service can handle both IPv4 and IPv6 workloads. Dual-stack load balancing is possible between services running any combination of IPv4 and IPv6.
+
+The Service API now has new fields to support dual-stack, replacing the single ipFamily field.
+* You can select your choice of IP family by setting `ipFamilyPolicy` to one of three options: SingleStack, PreferDualStack, or RequireDualStack. A service can be changed between single-stack and dual-stack (within some limits).
+* Setting `ipFamilies` to a list of families assigned allows you to set the order of families used.
+* `clusterIPs` is inclusive of the previous `clusterIP` but allows for multiple entries, so it’s no longer necessary to run duplicate services, one in each of the two IP families. Instead, you can assign cluster IP addresses in both IP families. Note that for a given pod, there is no possibility of setting multiple IP addresses in the same family.
+
+
+## Default behavior remains single-stack
+
+Starting in 1.20, the underlying networking for Kubernetes has included dual-stack whether or not a cluster was configured with the feature flag to enable dual-stack. Now that the feature is stable in 1.23, the feature flag is removed.
```

Comment:
Hi @bridgetkromhout I was strolling through the [Features Gates page](https://kubernetes.io/docs/reference/command-line-tools-reference/feature-gates/) and I saw that `IPv6DualStack` was beta in 1.21.
Should this be 1.20 or 1.21?
```suggestion
Starting in 1.21, the underlying networking for Kubernetes has included dual-stack whether or not a cluster was configured with the feature flag to enable dual-stack. Now that the feature is stable in 1.23, the feature flag is removed.
```

Reply:
> Now that the feature is stable in 1.23, the feature flag is removed.

As this would go out after the release, we should use the past tense. For example:
```markdown
Kubernetes 1.23 removed that feature flag as part of graduating the feature to stable:
dual-stack networking is always available if you want to configure it.
You can set your cluster network to operate as single-stack IPv4, as single-stack IPv6,
or as dual-stack IPv4 / IPv6. Each of those modes is fully supported.
```

Reply:
> Hi @bridgetkromhout I was strolling through the [Features Gates page](https://kubernetes.io/docs/reference/command-line-tools-reference/feature-gates/) and I saw that `IPv6DualStack` was beta in 1.21. Should this be 1.20 or 1.21?

This statement is nuanced as dual stack code that enables nodes/services CIDRs to carry multiple address families without the k8s dual stack APIs being enabled  as been around since the reimplemented Alpha (when thinking about it conceptually from cluster configuration below the k8s API, then APIs to enable dual stack usage). It's probably worth rephrasing so that it's more clear.

Reply:
Okay, I tried rephrasing. Further edits welcome.